lib/runAll.js

```diff
@@ -107,14 +113,40 @@ export const runAll = async (
     return ctx
   }
 
-  const stagedFileChunks = chunkFiles({ baseDir: gitDir, files, maxArgLength, relative })
-  const chunkCount = stagedFileChunks.length
-  if (chunkCount > 1) debugLog(`Chunked staged files into ${chunkCount} part`, chunkCount)
+  const configDir2config = {
+    [gitDir]: config,
+  }
+  const configDir2Files = {
+    [gitDir]: [],
+  }
+  if (sparseConfig) {
+    for (const file of files) {
+      const absoluteFilepath = path.resolve(gitDir, file)
+      const subConfigInfo = await loadConfig({
+        searchStartPath: path.dirname(absoluteFilepath),
+        logger,
+      })
+      if (!subConfigInfo) {
+        debugLog(`${ConfigNotFoundError.message} for ${file}.`)
+        continue
+      }
+      const configDir = path.dirname(subConfigInfo.filepath)
+      if (!configDir2config[configDir]) {
+        configDir2config[configDir] = subConfigInfo.config
+      }
+      if (!configDir2Files[configDir]) {
+        configDir2Files[configDir] = []
+      }
+      configDir2Files[configDir].push(file)
```

Comment:
Would it make sense to always discover deeper config files and then print a warning if the `--sparse-config` mode is not used? I think this might make sense, for example:

```
⚠ Found multiple lint-staged config files but `--sparse-config` was not used.
  Only the topmost config file will be used:
    - .lintstagedrc.json
    - packages/a/.lintstagedrc.json
```

Reply:
> I wonder if we could do this always, also in the case of non-sparse mode. So if we move all the config loading here so that the default "root" config is always loaded, and then in the case of sparse mode more files are discovered.

Continue with this comment.

I think this is a good idea. but to be clarify, only `root` config file and **deepest** config file.

Assuming structure:

```
monorepo
  |- .lintstagedrc.json
  |- a/
      |- .lintstagedrc.json
      |- b/
          |- .lintstagedrc.json
          |- c/
             |- .lintstagedrc.json
             |- index.js
```

and git staged file is `a/b/c/index.js`

We should just warn about `.lintstagedrc.json` and `a/b/c/.lintstagedrc.json`. Becuase `a/.lintstagedrc.json` and `a/b/.lintstagedrc.json` is meaningless for `a/b/c/index.js` file

Reply:
After a second thought, it might hurt performance in current non sparse config. Like 100 files in git stage. Originally, user load "root" config once and run 100 tasks. if we provide this warning, 100 searching config logic needs to be run. Do you think it is necessary to accept this penality? @iiroj

Reply:
> After a second thought, it might hurt performance in current non sparse config. Like 100 files in git stage. Originally, user load "root" config once and run 100 tasks. if we provide this warning, 100 searching config logic needs to be run. Do you think it is necessary to accept this penality? @iiroj

Maybe i can lower the cost by cache config with configDir. In a general case, 100 stage files is like

```
packages/a/1.js
pacakges/a/2.js
packages/a/.....
packages/a/50.js

packages/b/1.js
packages/b/2.js
packages/b/...
packages/b/50.js
```

So, even 100 search logic runs. if we cache config by `packages/a` and `packages/b`. It can avoid lots of IO costs.

let me try it